python/taichi/lang/impl.py

```diff
@@ -832,6 +838,10 @@ def vars2entries(_vars):
                 if len(_var) > 0 and isinstance(
                         _var[0], str) and _var[0] == '__ti_format__':
                     res = _var[1:]
+                elif len(_var) > 0 and isinstance(
+                        _var[0], str) and _var[0] == '__ti_fmt_list__':
+                    yield _var[1:]
```

Comment:
put `res = _val[1:]` here so the whole function yields at the same place?

Reply:
I rewrote this part of code and use a tag named `'__ti_fmt_value__'` when referring to an expr-format pair (for example a `{x:.2f}` in the f-string).

`'__ti_format__'` is used to indicate an artifact list generated in translating `str.format()`, instead of a user inputted one. After removing this tag, the remaining part is still a list and should be handled recursively in `vars2entries()`, so rest part of the `_val` should be pass to `vars2entries()` again then.

On the other hand, `'__ti_fmt_value__'` indicates a single Expr with format, so I think we should create a new code path to yield Expr and format together.